src/uint/gcd.rs

```diff
@@ -1,18 +1,34 @@
-//! Support for computing greatest common divisor of two `Uint`s.
+//! Support for computing the greatest common divisor of two `Uint`s.
 
-use crate::{modular::BernsteinYangInverter, ConstCtOption, Gcd, Odd, PrecomputeInverter, Uint};
+use crate::{
+    modular::BernsteinYangInverter, ConstChoice, ConstCtOption, Gcd, Odd, PrecomputeInverter, Uint,
+};
 use subtle::CtOption;
 
 impl<const SAT_LIMBS: usize, const UNSAT_LIMBS: usize> Uint<SAT_LIMBS>
 where
     Odd<Self>: PrecomputeInverter<Inverter = BernsteinYangInverter<SAT_LIMBS, UNSAT_LIMBS>>,
 {
     /// Compute the greatest common divisor (GCD) of this number and another.
-    ///
-    /// Returns none in the event that `self` is even (i.e. `self` MUST be odd). However, `rhs` may be even.
     pub const fn gcd(&self, rhs: &Self) -> ConstCtOption<Self> {
-        let ret = <Odd<Self> as PrecomputeInverter>::Inverter::gcd(self, rhs);
-        ConstCtOption::new(ret, self.is_odd())
+        let k1 = self.trailing_zeros();
+        let k2 = rhs.trailing_zeros();
+
+        // Select the smaller of the two `k` values, making 2^k the common even divisor
+        let k = ConstChoice::from_u32_lt(k2, k1).select_u32(k1, k2);
+
+        // Decompose `self` and `rhs` into `s{1, 2} * 2^k` where either `s1` or `s2` is odd
+        let s1 = self.overflowing_shr(k).unwrap_or(Self::ZERO);
```

Comment:
Probably `wrapping_shr()` would look a little better here (even though it does the same thing), since `k` is always less than `BITS`.

Reply:
Ah actually it is not, it will be `BITS` if `self` and `rhs` are zero.